scipy/signal/_signaltools.py

```diff
@@ -2210,15 +2214,17 @@ def lfilter(b, a, x, axis=-1, zi=None):
                     elif k != axis and zi.shape[k] == 1:
                         strides[k] = 0
                     else:
-                        raise ValueError('Unexpected shape for zi: expected '
+                        raise ValueError('Unexpected shape for parameter zi: expected '
                                          f'{expected_shape}, found {zi.shape}.')
                 zi = np.lib.stride_tricks.as_strided(zi, expected_shape,
                                                      strides)
             inputs.append(zi)
         dtype = np.result_type(*inputs)
 
         if dtype.char not in 'fdgFDGO':
-            raise NotImplementedError(f"input type '{dtype}' not supported")
+            p_types = tuple(p_.dtype.char for p_ in inputs)
+            raise NotImplementedError("Parameter dtypes must be convertible to one of "
+                                      f"'fdgFDGO'! Given input dtypes are {p_types}.")
```

Comment:
Character codes are very obscure and shouldn't be used in error messages. And object dtype is best left out. I think it's hard to improve this error message, since the "can be coerced to a real or complex floating point dtype" isn't very understandable for a user. I'd lean towards not changing the current message.